File: src/config/ranks.ts
import type { Rank } from '../types'

export const RANK_CONFIG: Record<Rank, {
  level: number
  icon: string
  minCorrect: number
  minAccuracy: number
  minTotal: number
}> = {
  '吃瓜群众': { level: 1, icon: 'Sprout', minCorrect: 0, minAccuracy: 0, minTotal: 0 },
  '瓜田新手': { level: 2, icon: 'Leaf', minCorrect: 5, minAccuracy: 30, minTotal: 10 },
  '鉴瓜学徒': { level: 3, icon: 'Flower2', minCorrect: 20, minAccuracy: 40, minTotal: 30 },
  '瓜田侦探': { level: 4, icon: 'Award', minCorrect: 50, minAccuracy: 50, minTotal: 80 },
  '鉴瓜达人': { level: 5, icon: 'Star', minCorrect: 100, minAccuracy: 55, minTotal: 150 },
  '鉴瓜大师': { level: 6, icon: 'Trophy', minCorrect: 200, minAccuracy: 60, minTotal: 300 },
  '见微先知': { level: 7, icon: 'Diamond', minCorrect: 500, minAccuracy: 65, minTotal: 600 },
}

export const RANK_ORDER: Rank[] = [
  '吃瓜群众',
  '瓜田新手',
  '鉴瓜学徒',
  '瓜田侦探',
  '鉴瓜达人',
  '鉴瓜大师',
  '见微先知',
]

export function getRankProgress(user: { rank: Rank; correctGuesses: number; totalGuesses: number }) {
  const currentRank = user.rank
  const currentIndex = RANK_ORDER.indexOf(currentRank)
  const nextRank = currentIndex < RANK_ORDER.length - 1 ? RANK_ORDER[currentIndex + 1] : null

  if (!nextRank) {
    return { current: currentRank, next: null, progress: 100, remaining: 0 }
  }

  const currentConfig = RANK_CONFIG[currentRank]
  const nextConfig = RANK_CONFIG[nextRank]

  const minCorrect = currentConfig.minCorrect
  const targetCorrect = nextConfig.minCorrect
  const range = targetCorrect - minCorrect
  const current = user.correctGuesses - minCorrect
  const progress = Math.min(100, Math.max(0, Math.round((current / range) * 100)))
  const remaining = Math.max(0, targetCorrect - user.correctGuesses)

  return {
    current: currentRank,
    next: nextRank,
    progress,
    remaining,
  }
}

File: backend/ranks.py
"""段位配置"""
RANK_CONFIG = {
    "吃瓜群众": {"level": 0, "min_correct": 0, "min_accuracy": 0, "min_total": 0, "icon": "🌱"},
    "瓜田新手": {"level": 1, "min_correct": 10, "min_accuracy": 0.25, "min_total": 30, "icon": "🌿"},
    "鉴瓜学徒": {"level": 2, "min_correct": 30, "min_accuracy": 0.35, "min_total": 80, "icon": "🍀"},
    "瓜田侦探": {"level": 3, "min_correct": 80, "min_accuracy": 0.45, "min_total": 180, "icon": "🔍"},
    "鉴瓜达人": {"level": 4, "min_correct": 200, "min_accuracy": 0.50, "min_total": 400, "icon": "🎯"},
    "鉴瓜大师": {"level": 5, "min_correct": 400, "min_accuracy": 0.55, "min_total": 800, "icon": "🏆"},
    "见微先知": {"level": 6, "min_correct": 800, "min_accuracy": 0.60, "min_total": 1500, "icon": "✨"},
}

def calculate_rank(correct_guesses: int, total_guesses: int) -> str:
    if total_guesses == 0:
        return "吃瓜群众"
    accuracy = correct_guesses / total_guesses
    ranks = list(RANK_CONFIG.keys())
    for i in range(len(ranks) - 1, -1, -1):
        rank = ranks[i]
        config = RANK_CONFIG[rank]
        if (correct_guesses >= config["min_correct"] and 
            total_guesses >= config["min_total"] and 
            accuracy >= config["min_accuracy"]):
            return rank
    return "吃瓜群众"
